### ISSUE TITLE
Preprocess should check roi valid when using YUV420 format

### ISSUE THREAD

**Issue by b04505009** [openvinotoolkit/dlstreamer_gst#209]
I'm using 2 stage detection pipeline like

gvadetect1 -> gvatrack -> gvadetect2

and gvadetect2 inference from roi-list. And I use NV12 as frame format.

I found this pipeline always stuck at a certain frame, and the error message is 
```
0:00:52.848018504 245844 0x7f3bb8019700 ERROR             GVA_common openvino_image_inference.cpp:175:operator(): Inference request completion callback failed with InferenceEngine::StatusCode: -1
	GENERAL_ERROR
```

Then I tried using opencv as preprocess backend, and the error message is 
```
error: Failed to submit images to inference
	Failed while software frame preprocessing
		Failed during OpenCV image pre-processing
			OpenCV(4.5.2-openvino) ../opencv/modules/imgproc/src/color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cvtColor'
```
I guess the problem is there are some roi which width or height were 0, so I add gvametapublish before and after gvatrack.

Here are the result json file:
[beforetrack.zip](https://github.com/openvinotoolkit/dlstreamer_gst/files/6340373/beforetrack.zip)
[aftertrack.zip](https://github.com/openvinotoolkit/dlstreamer_gst/files/6340374/aftertrack.zip)

The point is at the final line of aftertrack.json:
```json
    {
        "objects": [
            {
                "detection": {
                    "bounding_box": {
                        "x_max": 0.9994791666666666,
                        "x_min": 0.9989583333333333,
                        "y_max": 0.4662037037037037,
                        "y_min": 0.37407407407407406
                    },
                    "confidence": 1.0,
                    "label": "license plate",
                    "label_id": 2147483647
                },
                "h": 199,
                "id": 34,
                "roi_type": "license plate",
                "w": 1,
                "x": 1918,
                "y": 808
            }
        ],
        "resolution": {
            "height": 2160,
            "width": 1920
        },
        "timestamp": 60533333334
    }
```
There is a roi with w=1, though it's still fine.

Then I found out there are some issues in preprocess
https://github.com/openvinotoolkit/dlstreamer_gst/blob/4a8b37a81b4f8362899d3521bcabe5ecaec97014/inference_backend/pre_proc/opencv_utils/opencv_utils.cpp#L74-L94
`src.width` is 1 but after taking closest smaller even integer, `width` becomes 0, which causes cvtColor failed.

I think preprocesser should check if roi size is legal.

Another unrelated questions:
Is it normal that there are some objects' label_id are 2147483647?

**Comment by mliu2020** [comment]
@b04505009,

Thanks for the report. Let me check with dev team.

I have further question on the investigation:
- You mentioned about NV12 for input, what if BGR as input, is there still issue?
- If you remove gvatrack, would the problem still exist?
- Do you know which gvadetect stage did the issue happens, on first one or second one?

Mark

**Comment by b04505009** [comment]
Hi @mliu2020,

Thanks for your quick response.

1. The error happens in second gvadetect because it has to preprocess on previous roi.
2. It's a bit hard to test that because I already write my code for NV12 only, but I guess BGR might not have this issue.
3. No, the problem will not exist, but it was because the problematic roi with width 1 is generate from gvatrack. But I think this is not related to gvatrack. If there''s a case that first gvadetect would output roi with width 1, then this issue will still happen.
4. As 1.

b04505009

**Comment by b04505009** [comment]
Hi @mliu2020,

I modified your cpp sample draw_face_attributes to reproduce this error.

My modification:
1. Change capfilter to use NV12 (line 299)
2. Add a pad_probe_callback on gvadetect (line 314~320), and this callback will add a roi with width=1and height=1 (line 237)
3. Remove original pad_probe_callback on gvatrack because this will cause segmentation fault when using NV12 (line 323~329)
4. Using opencv as gvainference preprocess backend (line 295) because the program will just stuck and no error message  when using IE as backend. I guess both backend have the problem.

Here is the sample you could try.
[draw_face_attributes.zip](https://github.com/openvinotoolkit/dlstreamer_gst/files/6347646/draw_face_attributes.zip)

b04505009

**Comment by mliu2020** [comment]
Thanks so much for the work,

My apology for not respond quickly since I was little bit busy recently.

Let me check your package and submit a bug if I confirmed.

Mark

**Comment by mliu2020** [comment]
@b04505009 thanks so much for the detailed coding,

sorry for the late response. I was on other project and find time now to run your code.

Based on my understanding, this modified the original draw_face_attributes sample, 
- move the callback to gvadetect 'src' pad.
- In the callback so it add roi size (0,0,1,1).

I have one question: in the real world, the error happens randomly. So it would make sense to generate the problematic ROI at random position and only lasts 1~2 frames, does this make sense?

It has the following error I got:
```
developer@developer-desktop:~/Downloads/issue_209$ ./build_and_run.sh 
......
[100%] Built target draw_face_attributes
PIPELINE: urisourcebin buffer-size=4096 uri=https://github.com/intel-iot-devkit/sample-videos/raw/master/head-pose-face-detection-female-and-male.mp4 ! decodebin ! videoconvert n-threads=4 ! videoscale n-threads=4  ! capsfilter caps="video/x-raw,format=NV12" ! gvadetect name=gvadetect model=/home/developer/intel/dl_streamer/models/intel/face-detection-adas-0001/FP32//face-detection-adas-0001.xml device=CPU batch-size=1 ! queue ! gvainference model=/home/developer/intel/dl_streamer/models/intel/age-gender-recognition-retail-0013/FP32//age-gender-recognition-retail-0013.xml device=CPU batch-size=1 inference-region=roi-list pre-process-backend=opencv ! queue ! gvainference model=/home/developer/intel/dl_streamer/models/intel/emotions-recognition-retail-0003/FP32//emotions-recognition-retail-0003.xml device=CPU batch-size=1 inference-region=roi-list pre-process-backend=opencv ! queue ! gvainference model=/home/developer/intel/dl_streamer/models/intel/facial-landmarks-35-adas-0002/FP32//facial-landmarks-35-adas-0002.xml device=CPU batch-size=1 inference-region=roi-list pre-process-backend=opencv ! queue ! gvainference model=/home/developer/intel/dl_streamer/models/intel/head-pose-estimation-adas-0001/FP32//head-pose-estimation-adas-0001.xml device=CPU batch-size=1 inference-region=roi-list pre-process-backend=opencv ! queue ! gvawatermark name=gvawatermark ! videoconvert n-threads=4 ! fpsdisplaysink video-sink=xvimagesink sync=false 
ERROR from element gvainference0: base_inference failed on frame processing
Debugging info: /root/gst-video-analytics/gst/inference_elements/base/inference_singleton.cpp(176): frame_to_base_inference (): /GstPipeline:pipeline0/GstGvaInference:gvainference0:
Failed to submit images to inference
Failed while software frame preprocessing
Failed during OpenCV image pre-processing
OpenCV(4.5.1-openvino) ../opencv/modules/imgproc/src/color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cvtColor'
^C
```
The execution was stalled and I have to manually interrupt it.

Is this aligned with you?

Mark

**Comment by b04505009** [comment]
Hi @mliu2020 

Yes it makes sense to generate ROI in that situation, and that's what I met in my application and your result is aligned with me.

**Comment by mliu2020** [comment]
Yes, I have confirmed the same error happens if I modified your callback so it generated the problematic ROI at certain number of frames for only one time.

Mark

**Comment by adranit** [comment]
fixed in 2021.4.1
